File: retrofit_support/src/main/java/com/mdgd/commons/retrofit_support/BasicNetwork.java
package com.mdgd.commons.retrofit_support;

import com.mdgd.commons.result.ICallback;

import retrofit2.Call;

/**
 * Created by Owner
 * on 05/02/2019.
 */
public class BasicNetwork {

    protected <T, X> void execAsync(Call<T> call, ICallback<X> callback) {
        execAsync(call, callback, null);
    }

    protected <T, X> void execAsync(Call<T> call, ICallback<X> callback, ITransformer<T, X> transform) {
        call.enqueue(new RetroCallbackImpl<>(callback, transform));
    }

    protected <T, X> void execSync(Call<T> call, ICallback<X> callback) {
        execSync(call, callback, null);
    }

    protected <T, X> void execSync(Call<T> call, ICallback<X> callback, ITransformer<T, X> transform) {
        final RetroCallbackImpl<T, X> callbackImpl = new RetroCallbackImpl<>(callback, transform);
        try {
            callbackImpl.onResponse(call, call.execute());
        } catch (Throwable e) {
            callbackImpl.onFailure(call, e);
        }
    }
}

File: retrofit_support/src/main/java/com/mdgd/commons/retrofit_support/BasicNetwork.kt
package com.mdgd.commons.retrofit_support

import retrofit2.Call

open class BasicNetwork {

    fun <T, X>execAsync(call: Call<T>, callback: ICallback<X>?, transform: ITransformer<T, X>? = null) {
        call.enqueue(RetroCallbackImpl(callback, transform))
    }

    fun <T, X>execSync(call: Call<T>, callback: ICallback<X>?, transform: ITransformer<T, X>? = null) {
        val callbackImpl = RetroCallbackImpl(callback, transform)
        try {
            callbackImpl.onResponse(call, call.execute())
        } catch (e: Throwable){
            callbackImpl.onFailure(call, e)
        }
    }
}
